presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PickTableLayout.java

```diff
@@ -90,14 +100,55 @@ private static boolean shouldRewriteTableLayout(TableScanNode source)
         @Override
         public Result apply(FilterNode filterNode, Captures captures, Context context)
         {
+            TableScanNode tableScan = captures.get(TABLE_SCAN);
+
             TableLayoutRewriter tableLayoutRewriter = new TableLayoutRewriter(metadata, context.getSession(), context.getSymbolAllocator(), context.getIdAllocator());
             PlanNode rewrittenTableScan = tableLayoutRewriter.planTableScan(captures.get(TABLE_SCAN), filterNode.getPredicate());
 
-            if (rewrittenTableScan instanceof TableScanNode || rewrittenTableScan instanceof ValuesNode || (((FilterNode) rewrittenTableScan).getPredicate() != filterNode.getPredicate())) {
-                return Result.ofPlanNode(rewrittenTableScan);
+            if (arePlansEquivalent(filterNode, tableScan, rewrittenTableScan, context)) {
+                return Result.empty();
+            }
+
+            return Result.ofPlanNode(rewrittenTableScan);
+        }
+
+        public boolean arePlansEquivalent(FilterNode filter, TableScanNode tableScan, PlanNode rewritten, Context context)
+        {
+            Expression resultPredicate = BooleanLiteral.TRUE_LITERAL;
+            if (rewritten instanceof FilterNode) {
+                resultPredicate = ((FilterNode) rewritten).getPredicate();
+            }
+
+            Optional<TableScanNode> rewrittenTableScan = searchFrom(rewritten)
+                    .where(TableScanNode.class::isInstance)
+                    .findSingle();
+
+            if (!areExpressionEquivalent(filter.getPredicate(), resultPredicate, context)) {
+                return false;
             }
 
-            return Result.empty();
+            if (!rewrittenTableScan.isPresent()) {
+                return false;
+            }
+
+            return haveEqualConstraint(rewrittenTableScan.get(), tableScan);
+        }
+
+        private boolean areExpressionEquivalent(Expression first, Expression second, Context context)
+        {
+            if (ImmutableSet.copyOf(extractConjuncts(first)).equals(ImmutableSet.copyOf(extractConjuncts(second)))) {
+                return true;
+            }
```

Comment:
Why `equals(first, second)` is not sufficient (in practice)?
Why `extractConjuncts(first) eq extractConjuncts(second)` is not sufficient (in practice)?

Of course what you wrote is better, but it looks like this method needs to be as elaborate as required and is still might not be super future-proof, mightn't it?

Reply:
good question. I think with the code above we save one rule iteration. I will simplify that as it is not the point to save one point.